httpbakery/agent/agent.go

```diff
@@ -0,0 +1,181 @@
+// Package agent enables non-interactive (agent) login using macaroons.
+// To enable agent authorization with a given httpbakery.Client c against
+// a given third party discharge server URL u:
+//
+// 	SetUpAuth(c, u, agentUsername)
+//
+package agent
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"mime"
+	"net/http"
+	"net/url"
+
+	"github.com/juju/loggo"
+	"gopkg.in/errgo.v1"
+
+	"gopkg.in/macaroon-bakery.v1/bakery"
+	"gopkg.in/macaroon-bakery.v1/httpbakery"
+)
+
+var logger = loggo.GetLogger("httpbakery.agent")
+
+// agentLogin defines the structure of an agent login cookie. It is also
+// returned in a successful agent login attempt to help indicate that an
+// agent login has occurred.
+type agentLogin struct {
+	Username  string            `json:"username"`
+	PublicKey *bakery.PublicKey `json:"public_key"`
+}
+
+// agentResponse defines the structure of an agent login response.
+type agentResponse struct {
+	AgentLogin bool `json:"agent_login"`
+}
+
+// SetUpAuth configures agent authentication on c. A cookie is created in
+// c's cookie jar containing credentials derived from the username and
+// c.Key. c.VisitWebPage is set to VisitWebPage(c). The return is
+// non-nil only if c.Key is nil.
+func SetUpAuth(c *httpbakery.Client, u *url.URL, username string) error {
+	if c.Key == nil {
+		return errgo.New("cannot set-up authentication: client key not configured")
+	}
+	SetCookie(c.Jar, u, username, &c.Key.Public)
+	c.VisitWebPage = VisitWebPage(c)
+	return nil
+}
+
+// SetCookie creates a cookie in jar which is suitable for performing agent
+// logins to u.
+func SetCookie(jar http.CookieJar, u *url.URL, username string, pk *bakery.PublicKey) {
+	al := agentLogin{
+		Username:  username,
+		PublicKey: pk,
+	}
+	b, err := json.Marshal(al)
+	if err != nil {
+		// This shouldn't happen as the agentLogin type has to be marshalable.
+		panic(errgo.Notef(err, "cannot marshal cookie"))
+	}
+	v := base64.StdEncoding.EncodeToString(b)
+	jar.SetCookies(u, []*http.Cookie{{
+		Name:  "agent-login",
+		Value: v,
+	}})
+}
+
+// VisitWebPage creates a function that can be used with
+// httpbakery.Client.VisitWebPage. The function uses c to access the
+// visit URL which should previously have been configured with SetCookie.
```

Comment:
Whether the visit URL is or is not configured with SetCookie is not under the
control of the caller - it depends on the macaroon being discharged.

That makes me wonder if the function should check that there is
actually an agent-login cookie registered for the given URL before
doing the visit. What do you think?